### PR TITLE
Make OpenApp a write command

### PR DESCRIPTION
**Pull request by youssef-lr**
cc @iwiznia @tgolen @roryabraham @marcochavezf 

### Details

This PR makes OpenApp a write command so that for anonymous users we make sure to process `OpenApp` first before any requests that are made after it.

Summary of my findings from this [thread](https://expensify.slack.com/archives/C066HJM2CAZ/p1708134160129379)

> 1. The first request an anonymous user makes is OpenReport of the parent room, OpenReport is processed first and does not return a lastUpdateID
> 2. The app then makes a call to OpenApp as usual, but its response is not processed immediately as it’s a read command, the sequential queue must process write commands first, and in this case it’s still processing OpenReport
> 3. Anon clicks quickly on a thread and triggers another call to OpenReport which we will process before OpenApp ’s response
> 4. Once we [attempt](https://github.com/Expensify/App/blob/5b265f38d97972f0118ef4e3f08e620499b54580/src/libs/Middleware/SaveResponseInOnyx.ts#L45) to save the response of the 2ndOpenReport , the sequential queue finds out we don’t have a lastUpdateID set [here](https://github.com/Expensify/App/blob/5b265f38d97972f0118ef4e3f08e620499b54580/src/libs/actions/OnyxUpdateManager.ts#L71) and calls ReconnectApp
> 5. The sequential queue is paused until we process ReconnectApp which comes back with a lastUpdateID equal to OpenReport ’s
> 6. We unpause the queue to apply the update from OpenReport but since it has the same update ID we now have in the client we don’t apply its response and we end up with the component not having a report set.
> 
> So in conclusion, I think this is due to navigating to the thread quickly before OpenApp has finished, so if we make it a write, it’ll be processed first before we process  OpenReport when opening the thread 

### Fixed Issues

$ https://github.com/Expensify/App/issues/36430

### Tests

- [ ] Verify that no errors appear in the JS console

1. To reproduce locally add a `sleep(3)` to `openApp` [here](https://github.com/Expensify/Web-Expensify/blob/f8dc0034d2ab28d5efe4cdc91b139b71107178b3/lib/AppInit.php#L431), as to reproduce this bug `OpenReport` needs to finish before `OpenApp`.
2. Create a workspace room with visibility set to public, create a thread inside the room.
3. Open the room as an anonymous user in an incognito window, and click on the thread as fast as you can.
4. Make sure it loads properly.

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [X] I linked the correct issue in the `### Fixed Issues` section above
- [X] I wrote clear testing steps that cover the changes made in this PR
    - [X] I added steps for local testing in the `Tests` section
    - [X] I added steps for the expected offline behavior in the `Offline steps` section
    - [X] I added steps for Staging and/or Production testing in the `QA steps` section
    - [X] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [X] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I ran the tests on **all platforms** & verified they passed on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [X] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [X] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] I verified that if a function's arguments changed that all usages have also been updated correctly
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/9680864/dafc9c93-f2f9-4ed2-8914-e40da05007b2

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by marcochavezf** [comment]
Seems we'd need to update the unit test cases and types too

**Comment by youssef-lr** [comment]
Yeah I'm looking into it now

**Review by iwiznia — DISMISSED** [review]

**Review by tgolen — DISMISSED** [review]
In theory, I am not a fan of this change. I think it blurs the lines between why something is a read/write in the first place. If something is being made a read/write only to have some kind of effect in the UI, and it is not based on the command actually "reading or writing" to the database, then I think it's an indication that we are doing something wrong.

Since it sounds like the real intention is to have OpenApp go into the synchronous blocking queue, I'd be much more in favor of us opening up an option for that so that we can be explicit that this is still a READ command, but the request needs to go into the synchronous queue for X reason.

**Comment by youssef-lr** [comment]
@tgolen I agree and I thought about that, but `OpenApp` can actually be a write command as it creates a chat report [here](https://github.com/Expensify/Auth/blob/c1dee68835e9d5b26eba6de149a3e8301f6b8f4f/auth/command/OpenApp.cpp#L125) as @marcochavezf pointed out in the thread. Also, in Web-E `openApp` eventually calls `reconnectApp`, which is a write command, so I'm not really sure why we're having it as a read command.

**Comment by tgolen** [comment]
OK, thanks! That's great if it results in WRITEs being done.

**Comment by iwiznia** [comment]
>  Web-E openApp eventually calls reconnectApp, which is a write command

reconnectApp is a write command??? What does it write?

**Comment by youssef-lr** [comment]
>reconnectApp is a write command??? What does it write?

I'm not sure, if it doesn't actually write anything, it seems like we made it solely to have it block the queue? cc @tgolen I can see it was added [here](https://github.com/Expensify/App/pull/23516)

**Comment by tgolen** [comment]
The design doc said:

"It will call `API.write` (doing this so that it blocks all other calls)"

So yeah, I think your assumption is correct. No wonder why I didn't like
manipulating read/write like that!

File: tests/unit/NetworkTest.js [hunk 1]
```diff
@@ -168,17 +169,17 @@ describe('NetworkTests', () => {
                 // And then make 3 API requests in quick succession with an expired authToken and handle the response
                 // It doesn't matter which requests these are really as all the response is mocked we just want to see
                 // that we get re-authenticated
-                App.openApp();
-                App.openApp();
-                App.openApp();
+                PersonalDetails.openPersonalDetails();
```
Comment (youssef-lr) [inline tests/unit/NetworkTest.js:172]: I had to use a READ command here to make the test pass. It seems like this behaves differently when we call a write command.

Comment (youssef-lr) [reply tests/unit/NetworkTest.js:170]: I tested this scenario though of a write command with an expired authToken and I was re-authenticated properly.

**Review by iwiznia — DISMISSED** [review]

**Review by tgolen — CHANGES_REQUESTED** [review]

File: tests/unit/NetworkTest.js (hunk 1 above)
Comment (youssef-lr) [reply tests/unit/NetworkTest.js:170]: On it.

**Comment by youssef-lr** [comment]
cc'ing @marcaaron as well

**Review by marcaaron — APPROVED** [review]
These changes seem fine as far as I understand them. The "skew" stuff I am not sure about tbh still need to catch up on why we added that.

**Comment by youssef-lr** [comment]
Looks like we just need a checklist, since this is hard to reproduce on mobile, I think we can just do web screenshots.

**Review by tgolen — APPROVED** [review]

**Review by iwiznia — APPROVED** [review]

**Review by marcochavezf — APPROVED** [review]

**Pull request merged**